File: yologram-api-v1/src/main/kotlin/link/yologram/api/v1/domain/ums/service/AuthService.kt
package link.yologram.api.v1.domain.ums.service

import link.yologram.api.v1.domain.ums.exception.AuthWrongPasswordException
import link.yologram.api.v1.domain.ums.exception.UserNotFoundException
import link.yologram.api.v1.domain.ums.model.LoginRequest
import link.yologram.api.v1.domain.ums.model.LoginResponse
import link.yologram.api.v1.domain.ums.model.ValidateTokenResponse
import link.yologram.api.v1.domain.ums.repository.UserRepository
import link.yologram.api.v1.domain.ums.util.JwtUtil
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class AuthService(
    private val userRepository: UserRepository,
    private val passwordEncoder: BCryptPasswordEncoder,
    private val jwtUtil: JwtUtil,
) {

    fun login(request: LoginRequest): LoginResponse {
        val user = userRepository.findByEmail(request.email)
            .orElseThrow { UserNotFoundException() }

        if (!passwordEncoder.matches(request.password, user.password)) {
            throw AuthWrongPasswordException()
        }

        val accessToken = jwtUtil.createToken(user.id)

        return LoginResponse(
            uid = user.id,
            accessToken = accessToken,
            email = user.email,
            name = user.name,
            nickname = user.nickname,
        )
    }

    @Transactional(readOnly = true)
    fun validateToken(token: String): ValidateTokenResponse {
        val uid = jwtUtil.validateAndGetUid(token)
        val user = userRepository.findById(uid)
            .orElseThrow { UserNotFoundException() }

        return ValidateTokenResponse(
            uid = user.id,
            email = user.email,
            name = user.name,
            nickname = user.nickname,
        )
    }

    fun logout(uid: Long) {
    }
}

File: yologram-api-v2/app/domain/ums/auth_service.py
import bcrypt
from sqlalchemy.orm import Session

from app.core.exception import AuthWrongPasswordException, UserNotFoundException
from app.domain.ums.auth_schema import AuthData, LoginRequest, LoginResponse, ValidateTokenResponse
from app.domain.ums.jwt_util import create_token
from app.domain.ums.repository import UserRepository


class AuthService:

    def __init__(self, db: Session):
        self.repository = UserRepository(db)

    def login(self, request: LoginRequest) -> LoginResponse:
        user = self.repository.find_by_email(request.email)
        if not user:
            raise UserNotFoundException()

        if not bcrypt.checkpw(request.password.encode("utf-8"), user.password.encode("utf-8")):
            raise AuthWrongPasswordException()

        access_token = create_token(user.id)

        return LoginResponse(
            uid=user.id,
            access_token=access_token,
            email=user.email,
            name=user.name,
            nickname=user.nickname,
        )

    def validate_token(self, auth_data: AuthData) -> ValidateTokenResponse:
        user = self.repository.find_by_id(auth_data.uid)
        if not user:
            raise UserNotFoundException()

        return ValidateTokenResponse(
            uid=user.id,
            email=user.email,
            name=user.name,
            nickname=user.nickname,
        )

    def logout(self, auth_data: AuthData) -> None:
        pass
